hera_cal/abscal.py

```diff
@@ -3693,6 +3694,219 @@ def post_redcal_abscal_run(data_file, redcal_file, model_files, raw_auto_file=No
+                                   ref_antenna_name=refant_init, sky_catalog=f'{model_file}',
+                                   metadata_only=False, sky_field=field_str, cal_type='gain')
+    hc.write_calfits(output_filename, clobber=clobber)
+    hc = io.HERACal(output_filename)
+    hc.read()
+    hc.update(flags=data_ant_flags)
+    # generate cal object from model to hold model flags.
+    hcm = UVCal()
+    hcm = hcm.initialize_from_uvdata(uvdata=hdm, gain_convention='divide', cal_style='sky',
+                                     ref_antenna_name=refant_init, sky_catalog=f'{model_file}',
+                                     metadata_only=False, sky_field=field_str, cal_type='gain')
+    hcm.write_calfits(output_filename + '.tmp')
+    hcm = io.HERACal(output_filename + '.tmp')
+    hcm.read()
+    hcm.update(flags=model_ant_flags)
+    # init all gains to unity.
+    hcm.gain_array[:] = 1. + 0.j
+    hc.gain_array[:] = 1. + 0.j
+    # set calibration flags to be or of data and model flags
+    hc.flag_array = hc.flag_array | hcm.flag_array
+
+    abscal_gains, abscal_flags, _, _ = hc.build_calcontainers()
+
+    # load in weights if supplied.
+    if auto_file is not None:
+        hda = io.HERAData(auto_file)
+        bls = [ap for ap in hda.get_antpairs() if ap[0] == ap[1]]
+        auto_data, auto_flags, auto_nsamples = hda.read(bls=bls)
+        # generate wgts from autocorrelations
+        wgts = build_data_wgts(data_flags=data_flags, data_nsamples=data_nsamples, model_flags=model_flags,
+                               autocorrs=auto_data, auto_flags=auto_flags)
+    else:
+        # if no autocorrelation file supplied, use nsamples weights times
+        # or of data and model flags
+        wgts = DataContainer({k: (~data_flags[k]).astype(np.float64)
+                             * (~model_flags[k]).astype(np.float64)
+                             * data_nsamples[k] for k in data_flags})
+
+    # select refant if not specified.
+    if refant is None:
+        refant = pick_reference_antenna(abscal_gains, abscal_flags, hc.freqs, per_pol=True)
+    hc.ref_antenna_name = str(refant)
+    hcm.ref_antenna_name = str(refant)
+
+    abscal = AbsCal(data=data, model=model, wgts=wgts, freqs=data.freqs)
+
+    delta = 9e99
+
+    abscal_gains_iteration = {}
+
+    # find initial starting point with lincal before performing abscal.
+    if dly_lincal:
+        abscal.delay_lincal()
+        for k in abscal_gains:
+            abscal_gains_iteration[k] = abscal.ant_dly_phi_gain[k] * abscal.ant_dly_gain[k]
+        apply_cal.calibrate_in_place(data=abscal.data, new_gains=abscal_gains_iteration)
+        for k in abscal_gains:
+            abscal_gains[k] *= abscal_gains_iteration[k]
+
+    niter = 0
+    while delta > tol and niter < max_iter:
+        if not no_ampcal:
+            abscal.amp_logcal(verbose=verbose)
+        if not no_phscal:
+            abscal.phs_logcal(verbose=verbose)
+        for k in abscal.ant_eta_gain:
```

Comment:
this for loop will fail when no_ampcal is True but no_phscal is False b/c ant_eat_gain won't be populated.

you need two for loops, one for eta_gains and one for phi-gain. might as well put them both in the same conditional as the calibration step itself.

```
if not no_ampcal:
    abscal.amp_logcal()
    abscal_gains_iteration = abscal.merge_gains[(abscal_gains_iteration, abscal.ant_eta_gain]
```